### ISSUE TITLE
Training tips

### ISSUE THREAD

**Issue by BartvanMarrewijk** [IyatomiLab/LeafGAN#6]
Hi @huuquan1994 ,

I have trained cycle gan, leaf gan and a custom leaf gan using a detectron segmentator succesfully on the 2020 plant pathology [dataset](https://www.kaggle.com/c/plant-pathology-2020-fgvc7). I only used 2 classes of this dataset: healthy or scab. Using this data I trained resnet50 which was used as a benchmark. Then I training this resnet again but now with extra data from one of the gans (50% of train healthy images were converted to scab images). For validation the 2021 plant pathology dataset was used. 

Unfortunately, in none of the GAN datasets the classification accuracy improves. At first I thought, that some images were no transferred well. But even after removing the unclear images and training the classifier again the result was not yet better than the benchmark. 

Do you have any suggestions like training parameters? I know that in your paper you refer to the cycle gan paper, but even in that paper, the training tips are minimal. Did you change any parameters to get the result as described in your paper. Or do you think the added value of GAN is especially interesting when datasets are unbalanced?

**Comment by huuquan1994** [comment]
Hi @studentWUR,

Thanks for your question!
The key idea of LeafGAN is to increase the variety of backgrounds in training data. It depends on the diversity of the healthy input images (as we use it to transfer to fake diseases).
I've taken a look at the 2020 Plant Pathology Dataset. From my point of view, the healthy apple images seem to not have a wide variety of backgrounds and are quite similar to the disease's backgrounds. (Again, this is just my first impression since I didn't check the details on this dataset)
![image](https://user-images.githubusercontent.com/9111958/126854441-21eec560-90c9-4616-88ae-ca4a1a0569fa.png)
*healthy apple images taken from [this post](https://viblo.asia/p/kaggle-competition-plant-pathology-2020-fgvc7-phan-loai-benh-la-cay-oOVlYzBV58W)

Therefore, I'm not sure if LeafGAN works well on this dataset. In this case, I think if we can somehow segment the leaf area and "paste" into different backgrounds (not only apple but also like wild backgrounds) then the performance should be improved. Or maybe using more heavy augmentations [Albumtations](https://github.com/albumentations-team/albumentations) could help.

> Do you have any suggestions like training parameters? I know that in your paper you refer to the cycle gan paper, but even in that paper, the training tips are minimal. Did you change any parameters to get the result as described in your paper. Or do you think the added value of GAN is especially interesting when datasets are unbalanced?

In our experiments, we fine-tuned the pretrained ResNet-101 using the SGD optimizer with `lr=0.001` and `momentum=0.9` Here were the training hyperparameters.
```python
optimizer_ft = torch.optim.SGD(model_ft.parameters(), lr=0.001, momentum=0.9)
exp_lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer_ft, step_size=7, gamma=0.1)
```

Input image is normalized using mean and std of ImageNet dataset as:
```python
torchvision.transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
```

Also, the training datasets of our classifiers were balanced. Specifically,
* Baseline: Train on 4 diseases, 2000 images each.
* Baseline+CycleGAN/LeafGAN: Train on 4 diseases, 2000+717=2717 images each.

I hope this might help you and good luck with your model.

**Comment by BartvanMarrewijk** [comment]
Hi @huuquan1994 ,

Thank you for the elaborate answer. 

> Therefore, I'm not sure if LeafGAN works well on this dataset. In this case, I think if we can somehow segment the leaf area and "paste" into different backgrounds (not only apple but also like wild backgrounds) then the performance should be improved. Or maybe using more heavy augmentations Albumtations could help.

I fully agree that this could work, but then the higher performance is probably mainly caused by this augmentation instead of a GAN. My hypothesis was actually that by using LeafGAN we can generate examples that have the same background but the leaf itself is different; healthy or scab. Consequently, I expected that classifier should learn better, because the only differences between these images are the symptoms on the leaf. 

> Also, the training datasets of our classifiers were balanced. Specifically,
Baseline: Train on 4 diseases, 2000 images each.
Baseline+CycleGAN/LeafGAN: Train on 4 diseases, 2000+717=2717 images each.

Thank you for the clarification and training suggestions. Actually a nice test for the LeafGAN dataset is to determine the increase in performance when only the 717 healthy images were added to the baseline. In the table below, my results improved when I would compare with the original dataset _(Original (only fgvc7) VS GANs)_. However, when I added the 64 healthy images to original data set _(Original (only fgvc7) + 64 healthy)_. The results of the GAN are actually disappointing. Maybe I should have done the detectron GAN with more classes. And I agree with you when the background of the scab images would be different then the added value of LeafGAN would increase. 

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/marre004/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/marre004/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/marre004/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=NL style='tab-interval:35.4pt;word-wrap:break-word'>




Validation on fgvc8 (713, 670) | Train_size (healthy, scab) | Best.pt | Last.pt
-- | -- | -- | --
Original (only fgvc7) | 364 / 412 | 72.0 | 71.4
Original (only fgvc7) + 64 healthy | 418 / 412 | 77.3 | 71.8
Cycle gan +100% validA->B | 364 / 466 | 73.5 | 67.2
Leaf gan +100% validA->B | 364 / 466 | 73.02 | 68.18
Detectron gan+100% validA->B | 364 / 466 | 73.5 | 69.0




</body>

</html>

**Comment by soundaryabc123** [comment]
can you please update colab link. I'm finding difficulty in executing the code. I'm confused with what to run. Kindly explain it step by step it will help a students excelently.
Thank you
